pennylane/math/fidelity.py

```diff
@@ -0,0 +1,331 @@
+        )
+
+    return grad_dm0 * qml.math.cast_like(grad_out, grad_dm0)[:, None, None]
+
+
+def _compute_fidelity_vjp1(dm0, dm1, grad_out):
+    """
+    Compute the VJP of fidelity with respect to the second density matrix
+    """
+    # pylint: disable=arguments-out-of-order
+    return _compute_fidelity_vjp0(dm1, dm0, grad_out)
+
+
+def _compute_fidelity_grad(dm0, dm1, grad_out):
+    return _compute_fidelity_vjp0(dm0, dm1, grad_out), _compute_fidelity_vjp1(dm0, dm1, grad_out)
+
+
+################################ numpy ###################################
+
+ar.register_function("numpy", "compute_fidelity", _compute_fidelity_vanilla)
+
+################################ autograd ################################
+
+
+@autograd.extend.primitive
+def _compute_fidelity_autograd(dm0, dm1):
+    return _compute_fidelity_vanilla(dm0, dm1)
+
+
+def _compute_fidelity_autograd_vjp0(_, dm0, dm1):
+    def vjp(grad_out):
+        return _compute_fidelity_vjp0(dm0, dm1, grad_out)
+
+    return vjp
+
+
+def _compute_fidelity_autograd_vjp1(_, dm0, dm1):
+    def vjp(grad_out):
+        return _compute_fidelity_vjp1(dm0, dm1, grad_out)
+
+    return vjp
+
+
+autograd.extend.defvjp(
+    _compute_fidelity_autograd, _compute_fidelity_autograd_vjp0, _compute_fidelity_autograd_vjp1
+)
+ar.register_function("autograd", "compute_fidelity", _compute_fidelity_autograd)
+
+################################# jax #####################################
+
+try:
+    import jax
+
+    @jax.custom_vjp
+    def _compute_fidelity_jax(dm0, dm1):
+        return _compute_fidelity_vanilla(dm0, dm1)
+
+    def _compute_fidelity_jax_fwd(dm0, dm1):
+        fid = _compute_fidelity_jax(dm0, dm1)
+        return fid, (dm0, dm1)
+
+    def _compute_fidelity_jax_bwd(res, grad_out):
+        dm0, dm1 = res
+        return _compute_fidelity_grad(dm0, dm1, grad_out)
+
+    _compute_fidelity_jax.defvjp(_compute_fidelity_jax_fwd, _compute_fidelity_jax_bwd)
+    ar.register_function("jax", "compute_fidelity", _compute_fidelity_jax)
+
+except ModuleNotFoundError:
+    # jax not installed
+    pass
+
+################################ torch ###################################
+
+try:
+    import torch
+
+    class _TorchFidelity(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, dm0, dm1):
+            """Forward pass for _compute_fidelity"""
+            fid = _compute_fidelity_vanilla(dm0, dm1)
+            ctx.save_for_backward(dm0, dm1)
+            return fid
+
+        @staticmethod
+        def backward(ctx, grad_out):
+            """Backward pass for _compute_fidelity"""
+            dm0, dm1 = ctx.saved_tensors
+            return _compute_fidelity_grad(dm0, dm1, grad_out)
+
+    ar.register_function("torch", "compute_fidelity", _TorchFidelity.apply)
+
+except ModuleNotFoundError:
+    # torch not installed
+    pass
+
+############################### tensorflow ################################
+
+try:
+    import tensorflow as tf
```

Comment:
Did we have these try/except imports in the original file?

Reply:
No. These are used for registering the custom functions with autoray. If the import doesn't exist, then there's no need to register.

Reply:
Instead we should follow what is happening in `single_dispatch`, where the import only happens when needed.

Reply:
Changed to import at runtime 👍